qcengine/programs/tests/test_programs.py

```diff
@@ -200,3 +200,17 @@ def test_random_failure_with_success(failure_engine):
     assert ret.success, ret.error.error_message
     assert ret.provenance.retries == 1
     assert ret.extras["ncalls"] == 2
+
+@testing.using_openmm
+def test_openmm_task():
+    json_data = copy.deepcopy(_base_json)
+    json_data["molecule"] = qcng.get_molecule("water")
+    json_data["driver"] = "energy"
+    json_data["model"] = {"method": "smirnoff-1.1.0",
+                          "url": None,
+                          "offxml": "test_forcefields/smirnoff99Frosst.offxml"}
```

Comment:
Is the `offxml` field required here? or is `method` sufficient?

Reply:
How do we want this to work? 

What if we had `method` specify `smirnoff`, to distinguish this from adding support for the native OpenMM ffxml format later (or other formats in the future)?

Then we could have three possibilities:
* `filename` could specify a local path to a standard SMIRNOFF force field installed / distributed by the `openforcefield` toolkit
* `url` could specify a remote URL to a SMIRNOFF `.offxml` file
* `offxml` could specify the actual contents of a SMIRNOFF `offxml` file

Reply:
I like `smirnoff-1.1.0` as a full specifier to fully define a single method and can be supported in the database. The additional fields are useful for local computations I believe. A little messy would be `method:smirnoff + url or offxml`. If `Smirnoff-version` is specified url/offxml is ignore with warning?

Reply:
@dgasmith : The problem is that OpenMM is _way_ more general than just being able to compute energies with SMIRNOFF. In principle, we could add support for `gaff` versions, as well as `CGenFF`, or other small molecule force fields that can be computed in OpenMM. 

You could try to ask the `openforcefield` toolkit to resolve the parameter sets in order, and then fall back to something else (GAFF?) if not, but this logic gets complicated and there's no clear way to specify precendence.

We could _make_ this the "openforcefield" harness, rather than the OpenMM harness, but you'd be significantly limiting yourself for no clear reason.

Is there another clearer way we could specify "use the OpenMM engine and compute openforcefield energies with this force field" if you don't like the requirement of specifying either `filename`, `url`, or `offxml` when `"method" : "smirnoff"` is specified? This sounds simple, and we can have a simple precedence order, as well as warn if an unused property is specified.

Is the issue that QCArchive organizes data by `method`?

Reply:
Yes, it ultimately depends which level of abstraction you are looking for here. Quantum chemistry is pretty easy to generally restrict itself to a few specific field (`method`, `basis`, `keywords`) to fully define its method. Structured data like in QCArchive becomes a bit restricting as its hard to change primary keys. There are certainly ways around this, but outside our design scope at the moment.

> Is there another clearer way we could specify "use the OpenMM engine and compute openforcefield energies with this force field" if you don't like the requirement of specifying either filename, url, or offxml when "method" : "smirnoff" is specified? 

Yes, this is a better phrased version of what I was suggesting. 

QCArchive really only knows about `method` and `basis` fields here. We can either lump all the information into `method` or use `basis` somewhat awkwardly. Additional fields like `url` `filename` etc are of course perfectly fine here at the Engine layer, but will not cooperate with Fractal layers and above.

Reply:
@dotsdl A quick update here from talking to @jchodera  lets do the following:
1) `method` is always the typing engine (openff/openmm/etc)
2) `basis` is the forcefield name and version down to patch if available (Smirnoff-v1.0.0, UFF, etc)
3) `basis` need not be specified. If a `url` or `offxml` is supplied it will this and fill in the basis later. If the `url/offxml` leads to a non-versioned forcefield the `basis` name will be generated as `f"{forcefield_name}-{hash(forcefield_schema)}"`

I think this covers the bases.

Reply:
@dgasmith and @jchodera, is there something to grab onto in a SMIRNOFF XML file indicating `forcefield_name` and `forcefield_version`? For the `offxml` path case we can (somewhat reliably) use `os.path.splitext(os.path.basename(offxml))[0]`, but for the `url` case this may be (even) less reliable. Looking at the [`openff-1.0.0.offxml`](https://raw.githubusercontent.com/openforcefield/openforcefields/master/openforcefields/offxml/openff-1.0.0.offxml) content, I don't see something internal to the structure out of which I can pull this info for populating `basis`.

We could avoid this problem by demanding that `basis` be populated. The danger here of course is that `basis` may not actually match what is given by `offxml` or `url`. I'm not sure how likely this is to be a problem in practice, however.

Reply:
@j-wags is there a good way to get the openff xml version?

Reply:
Related question: given our use of pydantic, is it even possible to alter `model.basis` by the time it enters `OpenMMHarness.compute`?

Reply:
It cannot be changed, but it can be updated.